### ISSUE TITLE
Val loss increases when more of network is trained

### ISSUE THREAD

**Issue by ash1995** [matterport/Mask_RCNN#1259]
This is more of a question rather than an issue. I've noticed that while only training the heads of the network the training and validation loss decreases and plateaus after a certain number of epochs. When I start training parts of the backbone and eventually the entire network the model begins to overfit (validation loss increases while training loss decreases). 

My question is, should I stop trying to train the entire network and just train the network heads and stop once the losses stabilise? Or should I keep playing around with the hyperparameters to see if better performance can be obtained. 

I tried increasing the regularisation but it didn't seem to help once the model moved on to training parts of the backbone.

**Comment by K-M-Ibrahim-Khalilullah** [comment]
I trained MaskRCNN on my own dataset. During training, some values are look like this.

    Epoch 32/40
    200/200 [==============================] - 132s 660ms/step - loss: 0.1315 - rpn_class_loss: 9.4215e-04 - rpn_bbox_loss: 0.0107 - mrcnn_class_loss: 0.0074 - mrcnn_bbox_loss: 0.0088 - mrcnn_mask_loss: 0.1037 - val_loss: 1.6259 - val_rpn_class_loss: 0.1048 - val_rpn_bbox_loss: 0.5142 - val_mrcnn_class_loss: 0.2798 - val_mrcnn_bbox_loss: 0.1932 - val_mrcnn_mask_loss: 0.5338
    Epoch 33/40
    200/200 [==============================] - 133s 663ms/step - loss: 0.1308 - rpn_class_loss: 0.0012 - rpn_bbox_loss: 0.0111 - mrcnn_class_loss: 0.0069 - mrcnn_bbox_loss: 0.0090 - mrcnn_mask_loss: 0.1026 - val_loss: 1.3137 - val_rpn_class_loss: 0.0856 - val_rpn_bbox_loss: 0.4055 - val_mrcnn_class_loss: 0.2032 - val_mrcnn_bbox_loss: 0.1625 - val_mrcnn_mask_loss: 0.4570
    Epoch 34/40
    200/200 [==============================] - 133s 664ms/step - loss: 0.1240 - rpn_class_loss: 7.8069e-04 - rpn_bbox_loss: 0.0093 - mrcnn_class_loss: 0.0079 - mrcnn_bbox_loss: 0.0071 - mrcnn_mask_loss: 0.0989 - val_loss: 1.4614 - val_rpn_class_loss: 0.0831 - val_rpn_bbox_loss: 0.3221 - val_mrcnn_class_loss: 0.2878 - val_mrcnn_bbox_loss: 0.2030 - val_mrcnn_mask_loss: 0.5654

Why val_loss is greater than 1?
Which value will be better? How can I reduce?
Any suggestion please.
Thanks

**Comment by ashnair1** [comment]
Could be overfitting. Is the val loss always greater than train loss?

**Comment by K-M-Ibrahim-Khalilullah** [comment]
@ash1995 
Thanks for your comment. 
I think so, it could be overfitting. 
> Is the val loss always greater than train loss?

Yes, in this dataset it is always greater than loss!!
Previously I used different dataset, but that was OK.

However, How many ways or How can I remove this overfitting? Would you suggest me please?

**Comment by ashnair1** [comment]
You have two avenues to explore:

**Data**
One possibility is insufficient size of data. 
Try data augmentation. This repository supports the [imgaug](https://github.com/aleju/imgaug) library so check it out and apply as necessary. If your dataset is really small, try to get more data.

Check how the train-val split done. Check for class imbalance i.e. when a class A occurs more frequently than class B. I usually get a count of objects from my train and val sets to see if there are any problems. 

**Model**
Try training only the heads of the network. Try increasing regularisation.

You may have to conduct multiple experiments to find the right combination but hopefully this gives you an idea where to start.

**Comment by K-M-Ibrahim-Khalilullah** [comment]
@ash1995 
Thank you very much for your valuable suggestion.

**Comment by ashwin-999** [comment]
@ML-BOSS Hi, did increasing regularization help?
